File: scripts/data_gen/generate_data.py
import csv
import random
from faker import Faker

fake = Faker()

NUM_MENTORS = 10000
NUM_STUDENTS = 10000

# -------------------- DATA POOLS --------------------

domains = [
    "Software Engineering", "Data Science", "AI/ML",
    "Product Management", "UI/UX Design",
    "Finance", "Marketing", "Cybersecurity", "HR"
]

skills_map = {
    "Software Engineering": ["React", "Node.js", "System Design", "Java", "AWS"],
    "Data Science": ["Python", "Pandas", "ML", "Statistics"],
    "AI/ML": ["Deep Learning", "NLP", "Transformers", "TensorFlow"],
    "Product Management": ["Roadmaps", "Analytics", "Agile", "UX"],
    "UI/UX Design": ["Figma", "Wireframing", "User Research"],
    "Finance": ["Stocks", "Valuation", "Excel"],
    "Marketing": ["SEO", "Content", "Ads"],
    "Cybersecurity": ["Ethical Hacking", "Network Security"],
    "HR": ["Recruitment", "Soft Skills"]
}

companies = [
    "Google", "Amazon", "Microsoft", "Flipkart",
    "Zomato", "Infosys", "TCS", "Startup"
]

education_pools = [
    "B.Tech IIT", "M.Tech IIT", "MBA IIM",
    "B.Des NID", "B.Tech NIT", "Self-Taught"
]

availability_options = ["Weekends", "Evenings", "Flexible"]

languages_list = ["English", "Hindi", "English,Hindi"]

# -------------------- HELPERS --------------------

def get_skills(domain):
    pool = skills_map[domain]
    count = random.randint(2, 4)
    skills = random.sample(pool, min(count, len(pool)))
    return "|".join(skills)

# -------------------- GENERATORS --------------------

def generate_mentor(mentor_id):
    domain = random.choice(domains)
    return {
        "id": mentor_id,
        "name": fake.name(),
        "role": "Mentor",
        "experience_years": random.randint(3, 20),
        "domain": domain,
        "skills": get_skills(domain),
        "company": random.choice(companies),
        "designation": fake.job(),
        "location": fake.city(),
        "country": "India",
        "education": random.choice(education_pools),
        "availability": random.choice(availability_options),
        "price_per_hour": random.randint(15, 120),
        "rating": round(random.uniform(4.2, 5.0), 1),
        "total_sessions": random.randint(20, 500),
        "languages": random.choice(languages_list),
        "bio": fake.sentence(),
        "linkedin": f"https://linkedin.com/in/{fake.user_name()}",
        "verified": random.random() > 0.3,
        "profile_image": f"https://i.pravatar.cc/150?img={mentor_id % 70}"
    }

def generate_student(student_id):
    domain = random.choice(domains)
    return {
        "id": student_id,
        "name": fake.name(),
        "role": "Student",
        "current_level": random.choice(["Beginner", "Intermediate", "Advanced"]),
        "target_domain": domain,
        "interests": get_skills(domain),
        "current_skills": get_skills(domain),
        "experience_years": random.randint(0, 5),
        "education": random.choice(["B.Tech", "BCA", "MBA", "Student"]),
        "location": fake.city(),
        "country": "India",
        "goals": fake.sentence(),
        "preferred_availability": random.choice(availability_options),
        "budget_per_hour": random.randint(5, 50),
        "languages": random.choice(languages_list),
        "linkedin": f"https://linkedin.com/in/{fake.user_name()}",
        "profile_image": f"https://i.pravatar.cc/150?img={(student_id + 30) % 70}"
    }

# -------------------- MAIN --------------------

if __name__ == "__main__":
    print("Generating mentors...")
    mentors = [generate_mentor(i + 1) for i in range(NUM_MENTORS)]
    
    print("Saving mentors.csv...")
    with open("mentors.csv", "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=mentors[0].keys())
        writer.writeheader()
        writer.writerows(mentors)
        
    print("Generating students...")
    students = [generate_student(i + 1) for i in range(NUM_STUDENTS)]
    
    print("Saving students.csv...")
    with open("students.csv", "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=students[0].keys())
        writer.writeheader()
        writer.writerows(students)
        
    print("Done! Generated 10k mentors and 10k students.")

File: scripts/data_gen/generateData.js
const { faker } = require("@faker-js/faker");
const fs = require("fs");

const NUM_MENTORS = 10000;
const NUM_STUDENTS = 10000;

// -------------------- DATA POOLS --------------------

const domains = [
  "Software Engineering",
  "Data Science",
  "AI/ML",
  "Product Management",
  "UI/UX Design",
  "Finance",
  "Marketing",
  "Cybersecurity",
  "HR",
];

const skillsMap = {
  "Software Engineering": ["React", "Node.js", "System Design", "Java", "AWS"],
  "Data Science": ["Python", "Pandas", "ML", "Statistics"],
  "AI/ML": ["Deep Learning", "NLP", "Transformers", "TensorFlow"],
  "Product Management": ["Roadmaps", "Analytics", "Agile", "UX"],
  "UI/UX Design": ["Figma", "Wireframing", "User Research"],
  Finance: ["Stocks", "Valuation", "Excel"],
  Marketing: ["SEO", "Content", "Ads"],
  Cybersecurity: ["Ethical Hacking", "Network Security"],
  HR: ["Recruitment", "Soft Skills"],
};

const companies = [
  "Google",
  "Amazon",
  "Microsoft",
  "Flipkart",
  "Zomato",
  "Infosys",
  "TCS",
  "Startup",
];

const education = [
  "B.Tech IIT",
  "M.Tech IIT",
  "MBA IIM",
  "B.Des NID",
  "B.Tech NIT",
  "Self-Taught",
];

const availabilityOptions = ["Weekends", "Evenings", "Flexible"];

const languagesList = ["English", "Hindi", "English,Hindi"];

// -------------------- HELPERS --------------------

function getRandom(arr) {
  return arr[Math.floor(Math.random() * arr.length)];
}

function getSkills(domain) {
  return skillsMap[domain]
    .sort(() => 0.5 - Math.random())
    .slice(0, 3)
    .join("|"); // Using pipe to avoid CSV confusion
}

// -------------------- GENERATORS --------------------

function generateMentor(id) {
  const domain = getRandom(domains);

  return {
    id,
    name: faker.person.fullName(),
    role: "Mentor",
    experience_years: faker.number.int({ min: 3, max: 20 }),
    domain,
    skills: getSkills(domain),
    company: getRandom(companies),
    designation: faker.person.jobTitle(),
    location: faker.location.city(),
    country: "India",
    education: getRandom(education),
    availability: getRandom(availabilityOptions),
    price_per_hour: faker.number.int({ min: 15, max: 120 }),
    rating: (Math.random() * (5 - 4.2) + 4.2).toFixed(1),
    total_sessions: faker.number.int({ min: 20, max: 500 }),
    languages: getRandom(languagesList),
    bio: faker.lorem.sentence(),
    linkedin: `https://linkedin.com/in/${faker.internet.username()}`,
    verified: Math.random() > 0.3,
    profile_image: `https://i.pravatar.cc/150?img=${id % 70}`,
  };
}

function generateStudent(id) {
  const domain = getRandom(domains);

  return {
    id,
    name: faker.person.fullName(),
    role: "Student",
    current_level: getRandom(["Beginner", "Intermediate", "Advanced"]),
    target_domain: domain,
    interests: getSkills(domain),
    current_skills: getSkills(domain),
    experience_years: faker.number.int({ min: 0, max: 5 }),
    education: getRandom(["B.Tech", "BCA", "MBA", "Student"]),
    location: faker.location.city(),
    country: "India",
    goals: faker.lorem.sentence(),
    preferred_availability: getRandom(availabilityOptions),
    budget_per_hour: faker.number.int({ min: 5, max: 50 }),
    languages: getRandom(languagesList),
    linkedin: `https://linkedin.com/in/${faker.internet.username()}`,
    profile_image: `https://i.pravatar.cc/150?img=${(id + 30) % 70}`,
  };
}

// -------------------- CSV WRITER --------------------

function toCSV(data) {
  const headers = Object.keys(data[0]);
  const rows = data.map((obj) =>
    headers.map((h) => `"${String(obj[h]).replace(/"/g, '""')}"`).join(","),
  );
  return [headers.join(","), ...rows].join("\n");
}

// -------------------- GENERATE DATA --------------------

console.log("Generating mentors...");
const mentors = Array.from({ length: NUM_MENTORS }, (_, i) =>
  generateMentor(i + 1),
);

console.log("Generating students...");
const students = Array.from({ length: NUM_STUDENTS }, (_, i) =>
  generateStudent(i + 1),
);

// -------------------- SAVE FILES --------------------

fs.writeFileSync("mentors.csv", toCSV(mentors));
fs.writeFileSync("students.csv", toCSV(students));

console.log("✅ Data generated:");
console.log("mentors.csv (10K)");
console.log("students.csv (10K)");
